### ISSUE TITLE
Preview development mode warns on invalid JSON if excluding all logged in users from Analytics

### ISSUE THREAD

**Issue by LuckynaSan** [GoogleForCreators/web-stories-wp#7939]
## Bug Description

1.) [Story passes AMP validator](https://search.google.com/test/amp?id=u72qIF-C6TtMMeWXq6miOg), but Story's debug mode in the story preview is outputting the below warning when you enable the option to exclude all logged in users from Analytics in Site Kit by Google:

     INVALID_JSON_CDATA at 6:106
     The script tag contains invalid JSON that cannot be parsed

<img width="1228" alt="Screen Shot 2021-06-11 at 12 06 36 PM" src="https://user-images.githubusercontent.com/66372350/121724428-7d94f880-caad-11eb-9475-2b2e6864245f.png">

<img width="1260" alt="Screen Shot 2021-06-11 at 12 08 40 PM" src="https://user-images.githubusercontent.com/66372350/121725027-54289c80-caae-11eb-87aa-1cedbac2174a.png">

2.) Debug mode in the story preview is outputting inconsistent warnings on refresh on the same URL as above

https://user-images.githubusercontent.com/66372350/121724405-7241cd00-caad-11eb-8cf3-5858e3663333.mov

## Expected Behaviour
1.) The warning should not be present
2.) Debug results should be consistent. 

## Steps to Reproduce

1. Site Kit settings > edit Analytics > Exclude from Analytics all logged-in users
2. Preview story
3. See debug tab and notice invalid JSON error
4. Refresh page and notice inconsistent results

## Additional Context

related support topic: https://wordpress.org/support/topic/debug-error-issue-json/

- Plugin Version: 
Web Stories 1.7.2
Site Kit by Google 1.34.0

- WordPress Version:
- Operating System: macOS
- Browser: Chrome

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by swissspidy** [comment]
This would be a bug in Site Kit and should be reported there.

cc @felixarntz

**Comment by aaemnnosttv** [comment]
@swissspidy looks like a similar issue was raised in the AMP plugin before which changed the way it handled this:
https://github.com/ampproject/amp-wp/issues/4430

The validator extension also does not flag this as invalid:
![image](https://user-images.githubusercontent.com/1621608/122034046-cadcd700-cdd9-11eb-8543-32a2b43c8b87.png)

Is this something different about how Web Stories is validating the markup?

**Comment by swissspidy** [comment]
We use the same sanitizers, so there shouldn‘t be a difference. I can check next week though.

cc @westonruter for thoughts just in case

**Comment by westonruter** [comment]
I forget what the purpose is for this `__gaOptOutExtension` element. If it needs to be a JSON `script`, you could just give it the contents `{}`. 

Otherwise, if it just has to be some other tag that is allowed in the `head`, why not `meta`? You could add this instead:

```html
<meta id="__gaOptOutExtension">
```

**Comment by aaemnnosttv** [comment]
> I forget what the purpose is for this `__gaOptOutExtension` element.

@westonruter it is for opting out of all Google Analytics measurement in AMP.
 
> if it just has to be some other tag that is allowed in the `head`, why not `meta`? You could add this instead:
> 
> ```
> <meta id="__gaOptOutExtension">
> ```

This was what we [originally planned on doing](https://github.com/google/site-kit-wp/issues/1251#issuecomment-598918913), but this fails HTML validation without additional attributes.

> If it needs to be a JSON `script`, you could just give it the contents `{}`.

Yes, I found this passes the validation as well, but as far as I know the contents of the element aren't used at all, it simply relies on its presence.

The AMP validator does not flag the current empty element as invalid though and we also include AMP validation in our tests so that Site Kit does not generate invalid AMP markup. Is there another form of validation that Web Stories applies?

**Comment by westonruter** [comment]
> This was what we [originally planned on doing](https://github.com/google/site-kit-wp/issues/1251#issuecomment-598918913), but this fails HTML validation without additional attributes.

Then you could just add empty `name` and `content` attributes? Like so:

```html
<meta id="__gaOptOutExtension" name="sitekit-ga-opt-out" content="">
```

**Comment by westonruter** [comment]
> The AMP validator does not flag the current empty element as invalid though and we also include AMP validation in our tests so that Site Kit does not generate invalid AMP markup. Is there another form of validation that Web Stories applies?

A JSON+LD `script` with empty text content is not a validation error, but rather a validation _warning_. 

The Web Stories debugger doesn't seem to be differentiating between errors and warnings. But the AMP validator does:

![image](https://user-images.githubusercontent.com/134745/122292595-00052a00-ceab-11eb-8a9e-73c30f48f5b6.png)

![image](https://user-images.githubusercontent.com/134745/122292620-072c3800-ceab-11eb-8264-64f2355f348d.png)

**Comment by swissspidy** [comment]
Bit unfortunate that the Web Stories debugger doesn't mark it as a warning as it should. Definitely confusing for users.

@aaemnnosttv Are any of the suggested workarounds reasonable to add to Site Kit? e.g. adding `{}` contents.

In the meantime we could add some workaround in the Web Stories plugin for that.

**Comment by aaemnnosttv** [comment]
@swissspidy we've already [updated the element to use a `meta` tag instead](https://github.com/google/site-kit-wp/pull/3593) – sorry, I meant to follow up here 😅

There's still some debug output that looks similar to before but the errors are different and don't seem to be related to Site Kit which might have something to do with my environment or the fact that it's mostly an empty story 😄 

![image](https://user-images.githubusercontent.com/1621608/123077091-6695c500-d422-11eb-9789-cfc91d16400b.png)

**Comment by swissspidy** [comment]
@aaemnnosttv Awesome, thanks. Those other errors you see are unrelated and caused by:

* Previewing a draft (drafts don't have a rel=canonical in WP)
* Not adding a poster image
* Not adding a publisher logo

**Comment by swissspidy** [comment]
Closing this issue now as https://github.com/google/site-kit-wp/pull/3593 was merged and this will be fixed in Site Kit 1.36
